vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,391 @@
+	}()
+
+	return &ProxyApp{
+		Client:       rpc.NewClientWithCodec(codec),
+		terminate:    cancelContext,
+		onTerminated: onTerminated,
+		subProcess:   subProcess,
+	}, nil
+}
+
+func (proxy *ProxyApp) CreatePool(config string, debug bool) (int, error) {
+	var reply proxyrpc.CreatePoolResult
+	err := proxy.Call(
+		"ProxyVM.CreatePool",
+		proxyrpc.CreatePoolParams{
+			Debug: debug,
+			Param: config,
+		},
+		&reply)
+	if err != nil {
+		return 0, err
+	}
+
+	return reply.Count, nil
+}
+
+func (proxy *ProxyApp) CreateInstance(workdir string, index int) (vmimpl.Instance, error) {
+	var reply proxyrpc.CreateInstanceResult
+	err := proxy.Call(
+		"ProxyVM.CreateInstance",
+		proxyrpc.CreateInstanceParams{
+			Workdir: workdir,
+			Index:   index},
+		&reply)
+	if err != nil {
+		return nil, fmt.Errorf("failed to proxy.Call(\"ProxyVM.CreateInstance\"): %w", err)
+	}
+
+	atomic.AddInt32(&proxy.instancesCreated, 1)
+	return &instance{
+		ProxyApp: proxy,
+		ID:       reply.ID,
+	}, nil
+}
+
+type instance struct {
+	*ProxyApp
+	ID string
+}
+
+// Copy copies a hostSrc file into VM and returns file name in VM.
+// nolint: dupl
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	var reply proxyrpc.CopyResult
+	err := inst.ProxyApp.Call(
+		"ProxyVM.Copy",
+		proxyrpc.CopyParams{
+			ID:      inst.ID,
+			HostSrc: hostSrc,
+		},
+		&reply)
+	if err != nil {
+		return "", err
+	}
+
+	return reply.VMFileName, nil
+}
+
+// Forward sets up forwarding from within VM to the given tcp
+// port on the host and returns the address to use in VM.
+// nolint: dupl
+func (inst *instance) Forward(port int) (string, error) {
+	var reply proxyrpc.ForwardResult
+	err := inst.ProxyApp.Call(
+		"ProxyVM.Forward",
+		proxyrpc.ForwardParams{
+			ID:   inst.ID,
+			Port: port,
+		},
+		&reply)
+	if err != nil {
+		return "", err
+	}
+	return reply.ManagerAddress, nil
+}
+
+func buildMerger(names ...string) (*vmimpl.OutputMerger, []io.Writer) {
+	var wPipes []io.Writer
+	merger := vmimpl.NewOutputMerger(nil)
+	for _, name := range names {
+		rpipe, wpipe := io.Pipe()
+		wPipes = append(wPipes, wpipe)
+		merger.Add(name, rpipe)
+	}
+	return merger, wPipes
+}
+
+func (inst *instance) Run(
+	timeout time.Duration,
+	stop <-chan bool,
+	command string,
+) (<-chan []byte, <-chan error, error) {
+	merger, wPipes := buildMerger("stdout", "stderr", "console")
+	receivedStdoutChunks := wPipes[0]
+	receivedStderrChunks := wPipes[1]
+	receivedConsoleChunks := wPipes[2]
+	outc := merger.Output
+
+	var reply proxyrpc.RunStartReply
+	err := inst.ProxyApp.Call(
+		"ProxyVM.RunStart",
+		proxyrpc.RunStartParams{
+			ID:      inst.ID,
+			Command: command},
+		&reply)
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("error calling ProxyVM.Run with command %v: %w", command, err)
+	}
+
+	runID := reply.RunID
+	errc := make(chan error, 1)
+	timeoutSignal := time.After(timeout)
+
+	go func() {
+		defer func() {
+			inst.ProxyApp.Call(
+				"ProxyVM.RunStop",
```

Comment:
Currnetly `ProxyVM.RunStop` can be called after `ProxyVM.Close` b/c we do it after sending to the errc chan. Do we want to call them in deterministic order?

Reply:
Let's use it to support <-timeout and <-stop logic only.